src/engine/bufferscalers/enginebufferscalerubberband.cpp

```diff
@@ -17,9 +17,25 @@ using RubberBand::RubberBandStretcher;
 
 namespace {
 
+// TODO (XXX): this should be removed. It is only needed to work around
+// a Rubberband 1.3 bug.
 // This is the default increment from RubberBand 1.8.1.
 size_t kRubberBandBlockSize = 256;
 
+#define RUBBERBANDV3 (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
+
+// incredibly ugly solution, but I find this to be the least invasive.
```

Comment:
that is really to much magic. A good old `if()` in the using code does work and is way more readable IMHO:
```
if (IsEngineV2(m_pRubberBand)) {
... code for engine 2
} 
```
Is this a use case for the not_null ptr?

Reply:
right. KISS.

> Is this a use case for the not_null ptr?

kinda, though if I simplify the `IsEngineV2` and make it a private member function, it could even work without any args.

Reply:
done. Thanks for highlighting this obvious simplification.

Reply:
ot, though related, so I rather ask here instead of on [Zulip: Rubberband 3.0](https://mixxx.zulipchat.com/#narrow/stream/109171-development/topic/Rubberband.203.2E0):
regarding the preferences integration, I'll add `Rubberband R3 (experimental)` to the keylock engine combobox. This comes from the `KeylockEngine` enum in [enginebuffer.h](https://github.com/mixxxdj/mixxx/blob/d53e0b0a421f3f7f9fb9927944978b948db90eb4/src/engine/enginebuffer.h#L77-L81) (hidden by `#if (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)`).

If this item is selected and applied (CO changed) [`EngineBuffer::slotKeylockEngineChanged`](https://github.com/mixxxdj/mixxx/blob/d53e0b0a421f3f7f9fb9927944978b948db90eb4/src/engine/enginebuffer.cpp#L801) would call something like `EngineBufferScaleRubberBand::setEngineV3(bool)` to set an internal bool `isEngineV3` which in turn is used to set the RubberBand options.

does this sound reasonable?

Reply:
I'm not sure to be honest. I don't know how it is done internally but if it works for soundtouch, it should work for rubberband too. I'd also get rid of the `(experimental)` because R3 is quite stable, we just need to document that it takes even more CPU.

Somewhere you'll probably need to call RubberbandEngineScaler::onSampleRateChanged to construct the appropriate engine.

Reply:
> I'd also get rid of the (experimental) because R3 is quite stable

ah okay, I noticed the `-beta` tag but it's considered stable for a few days now.